src/org/rascalmpl/repl/rascal/RascalInterpreterREPL.java

```diff
@@ -132,14 +142,40 @@ public IDEServices initialize(Reader input, PrintWriter stdout, PrintWriter stde
             throw new IllegalStateException("Already initialized");
         }
         eval = buildEvaluator(input, stdout, stderr, services);
+
+        // Register watches for all watchable locations on the search path for automatic reloading
+        eval.getRascalResolver().collect().stream().filter(this::isWatchable).forEach(p -> {
+            try {
+                reg.watch(p, true, d -> sourceLocationChanged(p, d));
+            }
+            catch (IOException e) {
+                e.printStackTrace(stderr);
+            }
+        });
         return services;
     }
 
+    private boolean isWatchable(ISourceLocation loc) {
+        for (var capability : reg.capabilities(loc)) {
+            // see the `IOCapability` ADT in `IO.rsc`
+            switch (((IConstructor)capability).getName()) {
+                case "watching":
+                case "writing":
+                    return true;
+            }
+        }
+        return false;
+    }
 
     @Override
     public ICommandOutput handleInput(String command) throws InterruptedException, ParseError, StopREPLException {
         Objects.requireNonNull(eval, "Not initialized yet");
         synchronized(eval) {
+            Set<String> changes = new HashSet<>();
+            changes.addAll(dirtyModules);
+            dirtyModules.removeAll(changes);
+            eval.reloadModules(eval.getMonitor(), changes, URIUtil.rootLocation("reloader"));
```

Comment:
this should be moved inside of the try, to make sure we're not leaking extra exceptions.